runtime/pkg/pbutil/pbutil.go

```diff
@@ -95,32 +104,48 @@ func ToValue(v any) (*structpb.Value, error) {
 
 // ToStruct converts a map to a google.protobuf.Struct. It's similar to
 // structpb.NewStruct(), but it recurses on valToPB instead of structpb.NewValue
-// to add support for more types.
-func ToStruct(v map[string]any) (*structpb.Struct, error) {
+// to add support for more types. Providing t as a type hint is optional.
+func ToStruct(v map[string]any, t *runtimev1.StructType) (*structpb.Struct, error) {
 	x := &structpb.Struct{Fields: make(map[string]*structpb.Value, len(v))}
-	for k, v := range v {
-		if !utf8.ValidString(k) {
-			return nil, fmt.Errorf("invalid UTF-8 in string: %q", k)
+	if t == nil {
+		for k, v := range v {
+			if !utf8.ValidString(k) {
+				return nil, fmt.Errorf("invalid UTF-8 in string: %q", k)
+			}
+			var err error
+			x.Fields[k], err = ToValue(v, nil)
+			if err != nil {
+				return nil, err
+			}
 		}
-		var err error
-		x.Fields[k], err = ToValue(v)
-		if err != nil {
-			return nil, err
+	} else {
+		for _, f := range t.Fields {
```

Comment:
Is it guaranteed that the size of fields will always be equal to size of `v`. I see a check for size in `safeFieldType`

Reply:
It's generally the case, but not if fields have been removed from the schema (like in the time series APIs in this PR). 

The size check in `safeFieldType` is just to be extra safe.